### ISSUE TITLE
Feature Request: obtain user profile as a security check

### ISSUE THREAD

**Issue by omar-hijab** [abraunegg/onedrive#2362]
### Is your feature request related to a problem? Please describe.

I've deployed this onedrive client in my university. It works great. I would like to set up a webform where users enter the response URL and the webform back-end script then authorizes or re-authorizes onedrive. Of course, this is a huge security hole without some safeguards. 

### Describe the solution you'd like

Here is a solution that seems simple and effective:

_Upon obtaining a token and authorizing itself, the onedrive client then uses the token to access the user's profile.  If the profile email so obtained doesn't agree with the Linux account username running the onedrive client authorization, then the authorization is invalidated._

Notes

1. By checking the email, one also restricts to the organization/university.
2. The Webform backend runs as www-data, but is given sudo privileges to run as the account requesting onedrive authorization. How does it know which account that is? Not by asking through the webform directly, since anything from there is suspect, but from the profile obtained above.
3.  Since response URL is only valid for the user who generated it, the profile obtained by the above patch is guaranteed to be the correct profile.
4. Since the code already accesses driveUrl="https://graph.microsoft.com/v1.0/me/drive", this shouldn't be too hard, as the profile is under "https://graph.microsoft.com/v1.0/me".
5. To keep things simple, the onedrive client need only store the profile on disk. The rest can be done by a wrapper onedrive install script. Then the wrapper install script can add exceptions wherein it does not invalidate authorization.
6. I don't even know what language the source code is written in, so I wouldn't know where to start writing the patch. 

### Describe alternatives you've considered

i've tried various solutions, but I don't like any of them, so I've taken my webform down until there is a secure solution. The best solution is to avoid the university IT department entirely and add a small patch to onedrive.

**Comment by abraunegg** [comment]
@omar-hijab 
> The Webform backend runs as www-data, but is given sudo privileges to run as the account requesting onedrive authorization. How does it know which account that is? Not by asking through the webform directly, since anything from there is suspect, but from the profile obtained above.

Storing the account profile as a JSON or other, in an unencrypted manner so that it can be consumed by a process is not the right process here. That essentially is a security exposure in itself.

What I think you need to do is to investigate building Single Sign On (SSO) to your Web Form.

That way you are authenticating the user to the web form, and can use and trust those SSO credentials to validate that the user using the web form is the user that they say they are.

You could also implement 2FA here to further strengthen your security posture by sending a 6 digit OTP to their email address (as per their SSO details that was used to authenticate to the web form) and they have to then enter that OTP into the webform which further validates they are who they say they are before they transition onto the next step.

You can then use the `--auth-files` option where the client will present the authentication URL and wait to further authenticate the client for the response file to be present - for example:
```
--auth-files /path/to/where/my/auth/url/file/sits/authURLFile:/path/to/where/my/auth/response/file/sits/responseUrlFile
```

You can then read, using your web form the auth URL, even then use a Javascript process to open up a new window with that as the source, allowing the page to be opened. As you already have determined the URL is unique for that particular user.

The user can then authenticate, you can read the response URI, and then consume that as the response file part of `--auth-files`


> I don't even know what language the source code is written in, so I wouldn't know where to start writing the patch.

The GitHub repo clearly details what language has been used:

![image](https://user-images.githubusercontent.com/4956234/228944101-a2f61c8c-158d-42e2-93d4-c11ba710230d.png)

Please research: https://dlang.org/

**Comment by omar-hijab** [comment]
Hi,

Thanks for the feedback, it was very helpful. I simplified things by adding
the single line to onedrive.d:

std.file.write(cfg.accessTokenFilePath, accessToken);

right after the refreshToken is written to file. and a couple of supporting
lines in the other files.  The webform back-end script  uses this version
of the onedrive client  only once, to verify the user's identity. if valid,
I then move the refresh_token to their config folder, and do a resync as
them.  This way I avoid writing code from scratch to interact with the
microsoft servers, and instead rely on all the checks and precautions in
your code.

Since this integrates with my organization's identity management, SSO is
auto included. The webform

https://banl.newhaven.edu/computing/access/

is klutzy but at some point I'll add JS to hide the request-response loop.

Thanks again,
Omar

**Comment by abraunegg** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
